### PR TITLE
Add tests coverage for REST client reactive using HTTP2

### PR DESCRIPTION
**Pull request by jedla97**
### Summary

Adding test coverage for [QUARKUS-658](https://issues.redhat.com/browse/QUARKUS-658).

This cover basic scenarios which Quarkus users can use. 

The `http2ClientAsync` and `http2ClientSync` which are removed is covered by new tests. In addition the class these test were using were also removed. The ones from http-advanced were removed completely as Quarkus doesn't plan to support the non-reactive client.

Please select the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Dependency update
- [ ] Refactoring
- [ ] Backport
- [x] New scenario (non-breaking change which adds functionality)
- [ ] This change requires a documentation update
- [ ] This change requires execution against OCP (use `run tests` phrase in comment)

### Checklist:
- [x] Methods and classes used in PR scenarios are meaningful
- [x] Commits are well encapsulated and follow [the best practices](https://cbea.ms/git-commit/)

### CODE REVIEW

**Comment by jedla97** [comment]
The `HttpVersionAlpnWithConfigKeyIT.testAsyncResponseWithSingleClientOnSingleClient` and `HttpVersionAlpnWithConfigKeyIT.testSyncResponseWithSingleClientOnSingleClient` are expected to fail because the https://github.com/quarkusio/quarkus/issues/38514 but I PR for it is created and passing so it will be merged pretty soon.

Cookie test failing also on daily so it's not caused by this PR

**Comment by jcarranzan** [comment]
> The `HttpVersionAlpnWithConfigKeyIT.testAsyncResponseWithSingleClientOnSingleClient` and `HttpVersionAlpnWithConfigKeyIT.testSyncResponseWithSingleClientOnSingleClient` are expected to fail because the [quarkusio/quarkus#38514](https://github.com/quarkusio/quarkus/issues/38514) but I PR for it is created and passing so it will be merged pretty soon.
> 
> Cookie test failing also on daily so it's not caused by this PR

The issue is already closed so we can run again the tests

**Comment by jedla97** [comment]
The failing test are just cookies which will be fixed by https://github.com/quarkus-qe/quarkus-test-suite/pull/1639

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (jcarranzan) [inline http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java:85]: Consider using a constant for seconds (10).

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (jcarranzan) [inline http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java:139]: I think there is some method like `HttpVersion.HTTP_2.name()` it would be good to see it instead of the string "HTTP_2"

File: http/rest-client-reactive/src/test/java/io/quarkus/ts/http/restclient/reactive/HttpVersionAlpnWithConfigKeyIT.java
Comment (jcarranzan) [inline http/rest-client-reactive/src/test/java/io/quarkus/ts/http/restclient/reactive/HttpVersionAlpnWithConfigKeyIT.java:16]: Typo in  the word should be SUCCESSFUL_, you could change it in all SUCESSFUL_RESPONSE_STRING

**Review by jcarranzan — COMMENTED** [review]
Hi @jedla97, I've left some minor comments, but overall, the code looks good to me. Additionally, I was thinking that you might consider including a couple of tests that intentionally induce failures or check for the possibility of receiving a status code other than 200. 
Thank you

File: http/rest-client-reactive/src/main/resources/httpVersion.properties
Comment (jcarranzan) [inline http/rest-client-reactive/src/main/resources/httpVersion.properties:2]: According to the guide: https://es.quarkus.io/guides/http-reference#http2-support , the http2 property is already set by default, so you can remove it.

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (jedla97) [reply http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java:21]: What I know it doesn't matter it's always treated same. But I'll will add it as I see that is used everywhere else.

From docs of path
> For the purposes of absolutizing a path against the base URI , a leading '/' in a path is ignored and base URIs are treated as if they ended in '/'

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (jedla97) [reply http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java:69]: Thanks for the catch

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (jedla97) [reply http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java:39]: Good idea I change it

**Comment by jedla97** [comment]
Thanks for review @jcarranzan I update things which you have pointed out.

> Additionally, I was thinking that you might consider including a couple of tests that intentionally induce failures or check for the possibility of receiving a status code other than 200.

In this I will try thing about negative scenario and if I have some idea I'll create standalone PR for it.
In case of checking for 200 status in tests.

It's there yous to check but not testing the client itself. The test calling endpoint where is the client created/injected and that endpoint calling another endpoint as rest client. So status in tests is there just to check that's response is as expected (client use/not use http2). If there will be some problem with client it should return 500.

**Comment by jedla97** [comment]
Just note why we can't use the rest client directly in test:

https://github.com/quarkusio/quarkus/issues/25041

Tl;DR : At this moment quarkus rest client reactive is not meant to be used standalone.

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (jcarranzan) [reply http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java:72]: I keep seeing that, Have you already tweaked it in some way?

**Review by jcarranzan — COMMENTED** [review]
Thank you for the tweaks, just only a comment about the MalformedURLException signature and I'll approve it.

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (jedla97) [reply http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java]: Sorry I must forget to update it. Now it should be fine

File: http/rest-client-reactive/src/main/resources/httpVersion.properties
Comment (michalvavrik) [inline http/rest-client-reactive/src/main/resources/httpVersion.properties:1]: 1. properties comments shouldn't created with `//` whatever your framework works. If our framework really requires this, please open issue in Framework project
2. I don't think this comment is valuable. If you follow this pattern, you will have to comment out everything. One of million examples - `quarkus.http.ssl.certificate.key-store-file` in the file right above, is it better known to readers than following properties?
3. The comment doesn't seem true, this is client setting specific to concrete named clients?

I'm not against comments at all, but this one particular strikes me. Thouths @jedla97 ?

**Review by michalvavrik — COMMENTED** [review]
I don't have time to review this right now (not that I was asked to :-D), but once this is approved please do not merge. There is a lot of changes (nothing against them!) and I'd like to check when I find a time. Thanks

File: http/rest-client-reactive/src/main/resources/httpVersion.properties
Comment (jedla97) [reply http/rest-client-reactive/src/main/resources/httpVersion.properties]: Removed the comment. I had there part for server side settings so that make sense for me. But now agree that it's not needed.

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/HttpVersionClientWithConfigKey.java
Comment (jcarranzan) [inline http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/HttpVersionClientWithConfigKey.java:22]: Hi @jedla97 , I noticed some code duplication in the method interfaces between HttpVersionClientWithConfigKey and HttpVersionClient. Perhaps it could be improved by applying inheritance to a single interface containing the methods getClientHttpVersion and getClientHttpVersionAsync, and then having the rest of the interfaces extend from this "BaseInterface" with those methods.

Comment (jedla97) [reply http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/HttpVersionClientWithConfigKey.java]: @jcarranzan updated. I done it between the `ClientWithConfigKey` interfaces but not the rest. I agree that inherit it is better way. The `HttpVersionClientWithConfigKey` and `HttpsVersionClientWithConfigKey` inheriting from `HttpVersionClient` as they do same thing and are just specify name of `RestClient`. 

Personally don't want to introduce one new interface (base interface) and have 3 another interfaces inheriting from it. I see this base interface as something which are not needed + the inheritance from `HttpVersionClient` make sense for me.

Comment (jcarranzan) [reply http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/HttpVersionClientWithConfigKey.java]: Thank you now it make sense to me also :+1:

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (michalvavrik) [inline http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java:25]:
```suggestion
    public static final String WRONG_HTTP_VERSION = "The HTTP version should be HTTP/2 but is ";
```

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionBaseResource.java
Comment (michalvavrik) [inline http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionBaseResource.java:12]:
```suggestion
@Path("/http-version")
```
That's usual convention, or are you testing whether it works with case sensitive path? If so, let's document.

File: http/rest-client-reactive/src/test/java/io/quarkus/ts/http/restclient/reactive/HttpVersionAlpnWithConfigKeyIT.java
Comment (michalvavrik) [inline http/rest-client-reactive/src/test/java/io/quarkus/ts/http/restclient/reactive/HttpVersionAlpnWithConfigKeyIT.java:16]:
```suggestion
    private static final String SUCCESSFUL_RESPONSE_STRING = "Hello " + NAME;
```

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (michalvavrik) [inline http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java:131]: I think blocking here is just laziness, there is no point to it. Asynchronous test should be processed in a reactive manner and return stuff asynchronously from resource as well. Please feel free to prove me wrong. Thanks

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (michalvavrik) [inline http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java:120]:
```suggestion
     * Response from the client is checked if HTTP/2 was used
```
please check other comments as well. As you have already decided to write nice Java doc.

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionBaseResource.java
Comment (michalvavrik) [inline http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionBaseResource.java:18]: Hello is fine, you can keep it if you want as it tests query param. It would be nice to have incoming request protocol in response. This way, you can check that REST client is telling you a true about used HTTP version. I propose to inject `RoutingContext`, then `io.vertx.ext.web.RoutingContext#request` -> `io.vertx.core.http.HttpServerRequest#version` should always return `HTTP_2`.

**Review by michalvavrik — APPROVED** [review]
This looks fine. @jcarranzan please have a look next week, review and merge. If you need more time, can you maybe create dedicated ticket for yourself in JIRA, please? Let' s move on. Thxc

**Comment by michalvavrik** [comment]
@jedla97 let's add info about HTTP/2 for REST client into root project README

File: http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java
Comment (michalvavrik) [reply http/rest-client-reactive/src/main/java/io/quarkus/ts/http/restclient/reactive/resources/HttpVersionClientResource.java:25]: Now I'm thinking of it, you probably used actual response type value as they use underscore, fine then.

**Comment by jedla97** [comment]
@jcarranzan I update all the comments mentioned by @michalvavrik so can you have quick look. Also add mention of these test to the README.

**Comment by jcarranzan** [comment]
Changes LGTM, and tests are correct as well. However I noticed the test classes seem to contain repetitive tests, and perhaps using inheritance could consolidate tests into a single class. Nonetheless, I respect the current approach, it's simply a different way of approaching the tests. 
On the other hand, try to resolve the comments from Michal, still I can see some unresolved threads, thank you.

**Comment by jcarranzan** [comment]
Changes LGTM, and tests are correct as well. However I noticed the test classes seem to contain repetitive tests, and perhaps using inheritance could consolidate tests into a single class. Nonetheless, I respect the current approach, it's simply a different way of approaching the tests. 
On the other hand, try to resolve the comments from Michal, still I can see some unresolved threads, thank you.

**Comment by jedla97** [comment]
Thanks for review
> However I noticed the test classes seem to contain repetitive tests, and perhaps using inheritance could consolidate tests into a single class

When writing the test I was thinking about it, but if we want one class as base for multiple tests there would be subset of 2 tests for all 4 tests classes.
When we would have 2 base classes the subset of test would be 4 (all https tests). So in the result we shouldn't have 4 classes but 6. 

In my POV it's not that useful in this scenario.

> On the other hand, try to resolve the comments from Michal, still I can see some unresolved threads, thank you.

I update the solution but sum it in one comment as I was resolve all the comment on code side (not GitHub side)

**Pull request merged**
